manim/utils/tex.py

```diff
@@ -1,298 +1,179 @@
-"""Utilities for processing custom LaTeX templates."""
+"""Utilities for processing LaTeX templates."""
 
-__all__ = ["TexTemplateFromFile", "TexTemplate"]
+__all__ = [
+    "TexTemplate",
+    "TexTemplateFromFile",
+]
 
 
 import os
-from ..utils.config_ops import digest_config
 
 
-class TexTemplateFromFile:
-    """
-    Class representing a TeX template file
-    """  # TODO: attributes, dataclasses stuff
-
-    CONFIG = {
-        "use_ctex": False,
-        "filename": "tex_template.tex",
-        "text_to_replace": "YourTextHere",
-        "tex_compiler": {
-            "command": "latex",
-            "output_format": ".dvi",
-        },
-    }
-    body = ""
-
-    def __init__(self, **kwargs):
-        digest_config(self, kwargs)
-        self.rebuild_cache()
-
-    def rebuild_cache(self):
-        """For faster access, the LaTeX template's code is cached.
-        If the base file is modified, the cache needs to be rebuilt.
-        """
-        with open(self.filename, "r") as infile:
-            self.body = infile.read()
-
-    def get_text_for_text_mode(self, expression):
-        """Inserting expression verbatim into TeX template.
-
-        Parameters
-        ----------
-        expression : :class:`str`
-            String containing the expression to be typeset, e.g. ``"foo"``
-
-        Returns
-        -------
-        :class:`str`
-            LaTeX code based on the template containing the given expression and ready for typesetting.
-        """
-        return self.body.replace(self.text_to_replace, expression)
-
-    def get_text_for_env(self, environment, expression):
-        """Inserts an expression wrapped in a given environment into the TeX template.
-
-        Parameters
-        ----------
-        environment : :class:`str`
-            The environment in which we should wrap the expression.
-        expression : :class:`str`
-            The string containing the expression to be typeset, e.g. ``$\\sqrt{2}$``
-
-        Returns
-        -------
-        :class:`str`
-            LaTeX code based on template, containing the given expression and ready for typesetting
-        """
-        begin = r"\begin{" + environment + "}"
-        end = r"\end{" + environment + "}"
-        return self.body.replace(
-            self.text_to_replace, "{0}\n{1}\n{2}".format(begin, expression, end)
+class TexTemplate:
+    """TeX templates are used for creating Tex() and MathTex() objects."""
```

Comment:
please read our guidelines on class docstrings. This is lacking the attributes and stuff

Reply:
We should have this discussion sooner or later, and this reminds me: I would like to get rid of typing in the docstrings (while I do think that it improves documentation, it is **hard** to get right especially for people not into Sphinx/docstrings), and slowly introduce type hints (via PEP 484, https://www.python.org/dev/peps/pep-0484/). 

Sphinx is able to generate the same type information from type hints in the code as from the Numpy-style annotations in the docstring -- but IDEs can profit more from type hints.

What do you think, @PgBiel?

Reply:
The point of typing is to have docstrings be as precise as possible. And, most of what you'll need to write typings is in our guidelines.

Thing is, we do want to add typings; however, that requires people to contribute with typings on every PR, in order to not increase our workload indefinitely. So, whether we like it or not, people will need to write them. Enforcing typings on the docstring makes our job easier later. However, people may also write typehints and have us transport them to the docstring if they want an easier time (although they may do that by themselves).

So the original idea is to have both, in order to have things more organized.

Reply:
> Thing is, we do want to add typings; however, that requires people to contribute with typings on every PR, in order to not increase our workload indefinitely. 

I think it just has to be clear. Because in #457 just a few days ago, the position seemed to be different. If I may quote @leotrs:

> I think the typings can be left for a future PR. No need to request that right now, when we are not requesting the same from any other PR.
>
> In fact, there's an entirely different branch that is taking care of typings.

or @behackl:

> Yes, please, lets not overthink this. We don't have in-code typings anywhere in the code base yet, now and this PR is not the time and place to start introducing them.

Also, we should ask ourselves whether this might deter people from contributing valuable ideas, because they think they need to send "perfect" code. Adding typings can also be done later, it can be added as an easy issue. Anyhow, I will gladly help in adding needed typings.

Reply:
BTW: The Guidelines (https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP)#numpy-format) say: 

> `# typings are optional for now`

Reply:
To put context to that quote: someone had suggested we add typings on a PR that i) had been open for a while, and ii) it was suggested right at the last minute, once all other issues had been dealt with. At that point, my opinion was (and still is) that typings were not necessary to merge that PR.

Having said that, I think we should incorporate typings gradually, but we should not block PRs because they don't have typings. Usually I try to request some minimal documentation in each PR, but I think that filling in most sections of each docstring, as well as typings, can be left for the future.

I agree in general we should not turn away good/new ideas just because they lack typings.

Reply:
Thing is that, if we do plan on adding typings, then we need to start doing so now, because, otherwise, our workload will just increase indefinitely, as I said. Trust me, it's not that simple to start typing a repo with 0 types. So, by enforcing this, our job is way easier. By doing this, we're thinking of the future.

We don't have to add typehints in the code yet, but at least in the docstrings so that we know what to do later.

And regarding the guidelines: sorry if that wasn't clear; I meant that _typehints_ (i.e., in the code) are optional. Note the example that you quoted: it doesn't have any typings in its parameters, but it does have in the docstring. I'll update the guidelines to make that clear.

Reply:
I do know why typing is useful and I am not against having this kind of precision in our code base; far from it!

But at the same time, we have to be realistic. Adding Numpy-style docstrings, even though all of it is documented in our Wiki, is difficult to get right: the syntax is rather particular and unforgiving, and there is (at least in my IDE) no indication of whether the format I am writing is correct, leading to frustration for both contributors and reviewers.

This is why I would like to see us moving from typings in the docstrings to type hints, directly in the code (and most modern IDEs support checking your added type hints in some way, in particular whether they are syntactically correct). Adding these correctly is a lot easier (because the parser shouts at you) -- and thus also lowers the participation barrier for new (and existing) contributors.

I don't really see why you would like to have both type hints in the method signature and typings in the docstring: to be frank, I think wanting to have both is simply a disaster waiting to happen from a maintainance point of view. The rendered HTML documentation contains the same level of precision in both scenarios, and inspection shows PEP 484 type hints even better than typings added in the docstring.

Overall, I think type hints are easier to maintain than the docstring typings while, at the same time, yielding the same information. Whatever this discussion (that should probably be relayed to a new (or old?) issue) results in: I am strictly against maintaining typings in both the function signature via type hints and the docstring. I don't see how this makes sense or improves the available information in any way.

Reply:
Agreed we should start somewhere. However, this PR has been open for almost a month now so I'm not sure this is the right place to start doing so. It can be very annoying for a dev to be requested to add typings only after three weeks of working on this.

I think we should have a separate discussion about what to do with typings. Honestly, I think we can do without them, just like @behackl said, but @PgBiel has pushed hard for them in the past and I respect that.

Reply:
Thanks @PgBiel for the clarification. This leads me to another question: How and why did we get to the typing question here? It seems to me that the docstrings -- while lacking attrs at the class level --  contain all necessary typings. That's why I thought you were talking about typehints. (And because I was not aware of the subtle difference between these two terms.)

I will try to give @cobordism a hand for the class docstring.

Reply:
I started this, sorry. Let us take this to a dedicated issue. :-)

Reply:
After following the activity in this PR for a while I also think we need a better approach for this.

Reply:
@PhilippImhof yes, the docstrings should contain all necessary typings for now. Thing is, @behackl suggested we use typehints instead.

Here's my take on this: We can make typings on docs simpler. However, we may only do that _after_ we work on typings. Because we need at least _some way_ to deduce typings for parameters and stuff when we do it. Then, after said work is done, we may begin to enforce typehints for new PRs.

Also, note: You said that adding typehints makes the IDE parse and stuff. You can write them in the code, have the IDE suggest the changes and _then_ you add the typings to the docstring.

Reply:
Lets take this discussion to #52.

Reply:
@PgBiel 

> Here's my take on this: We can make typings on docs simpler. However, we may only do that _after_ we work on typings. 

I wouldn't want to be annoying, but are you talking about typings and typings or typings and typehints? I am just trying to use those terms correctly in the future.

> Also, note: You said that adding typehints makes the IDE parse and stuff. You can write them in the code, have the IDE suggest the changes and _then_ you add the typings to the docstring.

That wasn't me. 😄  My IDE is Emacs, it doesn't do anything like that. (It certainly could do it, but I do not have the necessary packages installed.)

Reply:
> I wouldn't want to be annoying, but are you talking about typings and typings or typings and typehints? I am just trying to use those terms correctly in the future.
Oh, I'm talking about typehints, yeah. And this part of the reply was meant for @behackl 

> That wasn't me.
Woops, sorry, I meant to ping @behackl 😅 sorry for that

Reply:
Any update on this? Documenting attrs and init params is necessary